src/evo/assetlocktx.cpp

```diff
@@ -0,0 +1,245 @@
+    // and at the quorumHash must be active in either the current or previous quorum cycle
+    // and the sig must validate against that specific quorumHash.
+
+    Consensus::LLMQType llmqType = Params().GetConsensus().llmqTypeAssetLocks;
+
+    if (!Params().HasLLMQ(llmqType)) {
+        return {ValidationInvalidReason::CONSENSUS, "bad-assetunlock-llmq-type"};
+    }
+
+    const auto& llmq_params = llmq::GetLLMQParams(llmqType);
+
+    // We check at most 2 quorums, so, count is equal to 2
+    // TODO: update it after choosing proper consensus.llmqTypeAssetLocks (see comment there)
+    const int count = 2;
+    auto quorums = llmq::quorumManager->ScanQuorums(llmqType, pindexTip, count > -1 ? count : llmq_params.signingActiveQuorumCount);
+    bool isActive = false;
+    for (const auto& q : quorums) {
+        if (q->qc->quorumHash == quorumHash) {
+            isActive = true;
+        }
+    }
+
+    if (!isActive) {
+        return {ValidationInvalidReason::CONSENSUS, "bad-assetunlock-not-active-quorum"};
+    }
+
+    if (pindexTip->nHeight < requestedHeight || pindexTip->nHeight >= getHeightToExpiry()) {
+        LogPrintf("Asset unlock tx %d with requested height %d could not be accepted on height: %d\n",
+                index, requestedHeight, pindexTip->nHeight);
+        return {ValidationInvalidReason::CONSENSUS, "bad-assetunlock-too-late"};
+    }
+
+    const auto quorum = llmq::quorumManager->GetQuorum(llmqType, quorumHash);
+    assert(quorum);
+
+    const std::string id(strprintf("plwdtx%lld", index));
+
+    std::vector<uint8_t> vchHash(32);
+    CSHA256().Write(reinterpret_cast<const uint8_t*>(id.data()), id.size()).Finalize(vchHash.data());
+    uint256 requestId(vchHash);
+
+    uint256 signHash = llmq::utils::BuildSignHash(llmqType, quorum->qc->quorumHash, requestId, msgHash);
+    if (quorumSig.VerifyInsecure(quorum->qc->quorumPublicKey, signHash)) {
+        return {};
+    }
+
+    return {ValidationInvalidReason::CONSENSUS, "bad-assetunlock-not-verified"};
+}
+
+maybe_error CheckAssetUnlockTx(const CTransaction& tx, const CBlockIndex* pindexPrev)
+{
+    if (tx.nType != TRANSACTION_ASSET_UNLOCK) {
+        return {ValidationInvalidReason::TX_BAD_SPECIAL, "bad-assetunlocktx-type"};
+    }
+
+    if (!tx.vin.empty()) {
+        return {ValidationInvalidReason::TX_BAD_SPECIAL, "bad-assetunlocktx-have-input"};
+    }
+
+    if (tx.vout.size() > CAssetUnlockPayload::MAXIMUM_WITHDRAWALS) {
+        return {ValidationInvalidReason::TX_BAD_SPECIAL, "bad-assetunlocktx-too-many-outs"};
+    }
+
+    CAssetUnlockPayload assetUnlockTx;
+    if (!GetTxPayload(tx, assetUnlockTx)) {
+        return {ValidationInvalidReason::TX_BAD_SPECIAL, "bad-assetunlocktx-payload"};
+    }
+
+    if (assetUnlockTx.getVersion() == 0 || assetUnlockTx.getVersion() > CAssetUnlockPayload::CURRENT_VERSION) {
+        return {ValidationInvalidReason::TX_BAD_SPECIAL, "bad-assetunlocktx-version"};
+    }
+
+    if (assetUnlockTx.getFee() <= 0) {
```

Comment:
Should we validate it against dust value?

Reply:
We should make sure we're not creating dust. This may be checked elsewhere already (e.g. for all transactions), but I'm not sure.

Reply:
this amount is summed up with other fees in block, so, overall fee output from all transaction won't be dust.

Reply:
Hmm, not sure I understand. It doesn't how many fees are in the block. There shouldn't be any output of any tx in the block below the dust amount. @PastaPastaPasta @UdjinM6 Will miners create blocks that contain dust outputs? Maybe this is a non-issue.

Reply:
I think what @knst is saying is that the asset lock fee is combined with the block reward, and as such is never dust